src/ILCompiler/src/Program.cs

```diff
@@ -333,33 +339,106 @@ private int Run(string[] args)
             else
                 builder = new RyuJitCompilationBuilder(typeSystemContext, compilationGroup);
 
+            ILScanResults scanResults = null;
+            if (_useScanner && !_isCppCodegen)
+            {
+                ILScannerBuilder scannerBuilder = builder.GetILScannerBuilder()
+                    .UseCompilationRoots(compilationRoots);
+
+                if (_scanDgmlLogFileName != null)
+                    scannerBuilder.UseDependencyTracking(_generateFullScanDgmlLog ? DependencyTrackingLevel.All : DependencyTrackingLevel.First);
+
+                IILScanner scanner = scannerBuilder.ToILScanner();
+
+                scanResults = scanner.Scan();
+            }
+
             var logger = _isVerbose ? new Logger(Console.Out, true) : Logger.Null;
 
             DependencyTrackingLevel trackingLevel = _dgmlLogFileName == null ?
                 DependencyTrackingLevel.None : (_generateFullDgmlLog ? DependencyTrackingLevel.All : DependencyTrackingLevel.First);
 
             CompilerGeneratedMetadataManager metadataManager = new CompilerGeneratedMetadataManager(compilationGroup, typeSystemContext, _metadataLogFileName);
 
-            ICompilation compilation = builder
+            builder
                 .UseBackendOptions(_codegenOptions)
                 .UseMetadataManager(metadataManager)
                 .UseLogger(logger)
                 .UseDependencyTracking(trackingLevel)
                 .UseCompilationRoots(compilationRoots)
                 .UseOptimizationMode(_optimizationMode)
-                .UseDebugInfo(_enableDebugInfo)
-                .ToCompilation();
+                .UseDebugInfo(_enableDebugInfo);
+
+            if (scanResults != null)
+                builder.UseVTableSliceProvider(scanResults.GetVTableLayoutInfo());
+
+            ICompilation compilation = builder.ToCompilation();
 
             ObjectDumper dumper = _mapFileName != null ? new ObjectDumper(_mapFileName) : null;
 
-            compilation.Compile(_outputFilePath, dumper);
+            CompilationResults compilationResults = compilation.Compile(_outputFilePath, dumper);
 
             if (_dgmlLogFileName != null)
-                compilation.WriteDependencyLog(_dgmlLogFileName);
+                compilationResults.WriteDependencyLog(_dgmlLogFileName);
+
+            if (scanResults != null)
+            {
+                SimdHelper simdHelper = new SimdHelper();
+
+                if (_scanDgmlLogFileName != null)
+                    scanResults.WriteDependencyLog(_scanDgmlLogFileName);
+
+                // If the scanner and compiler don't agree on what to compile, the outputs of the scanner might not actually be usable.
+                bool scanningFail = false;
+                DiffCompilationResults(ref scanningFail, compilationResults.CompiledMethodBodies, scanResults.CompiledMethodBodies,
```

Comment:
Needs comments to explain how the two sets of calls to DiffCompilationResults are tied together.